com.unity.shadergraph/Editor/Drawing/Controls/TextControl.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Linq;
+using System.Reflection;
+using UnityEngine;
+using UnityEngine.UIElements;
+using UnityEngine.Assertions;
+using UnityEditor.Graphing;
+
+namespace UnityEditor.ShaderGraph.Drawing.Controls
+{
+    [AttributeUsage(AttributeTargets.Property)]
+    class TextControlAttribute : Attribute, IControlAttribute
+    {
+        string m_Label;
+        public TextControlAttribute(string label = null)
+        {
+            m_Label = label;
+        }
+
+        public VisualElement InstantiateControl(AbstractMaterialNode node, PropertyInfo propertyInfo)
+        {
+            if (!TextControlView.validTypes.Contains(propertyInfo.PropertyType))
+                return null;
+            return new TextControlView(m_Label, node, propertyInfo);
+        }
+    }
+
+    class TextControlView : VisualElement
+    {
+        public static Type[] validTypes = { typeof(string) };
+        AbstractMaterialNode m_Node;
+        PropertyInfo m_PropertyInfo;
+        string m_Value;
+        int m_UndoGroup = -1;
+        public TextControlView(string label, AbstractMaterialNode node, PropertyInfo propertyInfo)
+        {
+            styleSheets.Add(Resources.Load<StyleSheet>("Styles/Controls/TextControlView"));
+            m_Node = node;
+            m_PropertyInfo = propertyInfo;
+            label = label ?? ObjectNames.NicifyVariableName(propertyInfo.Name);
+            var container = new VisualElement { name = "container" };
+            var thisLabel = new Label(label);
+            container.Add(thisLabel);
+            m_Value = GetValue();
+            string value = null;
+            var field = new TextField { value = m_Value };
+            field.RegisterCallback<MouseDownEvent>(Repaint);
+            field.RegisterCallback<MouseMoveEvent>(Repaint);
+            field.RegisterValueChangedCallback(evt =>
+            {
+                value = GetValue();
+                value = evt.newValue;
+            });
+
+            // Pressing escape while we are editing causes it to revert to the original value when we gained focus
+            field.Q("unity-text-input").RegisterCallback<KeyDownEvent>(evt =>
+            {
+                if (evt.keyCode == KeyCode.Escape && m_UndoGroup > -1)
+                {
+                    Undo.RevertAllDownToGroup(m_UndoGroup);
+                    m_UndoGroup = -1;
+                    evt.StopPropagation();
+                }
+                this.MarkDirtyRepaint();
+            });
+            field.Q("unity-text-input").RegisterCallback<FocusOutEvent>(evt =>
+            {
+                //Only set node value when mouse clicked away
+                m_Node.owner.owner.RegisterCompleteObjectUndo("Change" + m_Node.name);
+                m_PropertyInfo.SetValue(m_Node, value, null);
+                m_UndoGroup = -1;
+                //Validate graph to update downstream input slot
```

Comment:
If possible I'd make it so that there was an Action/Delegate here that triggered the responses of whichever node owns this TextControl. This way, every node can implement its own response to the value being changed. 

Might be a bit convoluted because this is all implemented as an Attribute, maybe could have another attribute for a method that defines the value change response? 

`TextControlValueChangedAttribute` or something like that maybe, just an idea, there are probably better ways to do it.

It's not a huge problem right now as this is only used by the SwizzleNode, but thinking about how this might need to scale to other nodes over time and all that.

Reply:
Had a discussion with Tracy offline and I think we're justified in keeping this in for now and looking at it later. The Swizzle node is the only user of the TextControlView in the graph at the moment and all of this code is likely to drastically change with GTF down the line so its not worth spending more time on what was a usability week task.